### ISSUE TITLE
Getting a self-signed certificate instead of LetsEncrypt one

### ISSUE THREAD

**Issue by elifiner** [JrCs/docker-letsencrypt-nginx-proxy-companion#543]
I'm running a simple `docker-compose.yml` configuration to test the setup and I'm getting a self signed certificate when accessing the service.

This is my `docker-compose.yml`:

```
version: '2'

services:
  nginx-proxy:
    image: jwilder/nginx-proxy:alpine
    ports:
      - "80:80"
      - "443:443"
    volumes:
      - /var/run/docker.sock:/tmp/docker.sock:ro
      - /etc/nginx/certs
      - /etc/nginx/vhost.d
      - /usr/share/nginx/html

  nginx-proxy-letsencrypt:
    image: jrcs/letsencrypt-nginx-proxy-companion
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock:ro
    volumes_from:
      - nginx-proxy

  whoami:
    image: jwilder/whoami
    environment:
      - VIRTUAL_HOST=whoami.mp0.rbgapp.com
      - LETSENCRYPT_HOST=whoami.mp0.rbgapp.com
```

This is the log when I run `docker-compose up`, which looks fine to me:

```
Creating network "proxy_default" with the default driver
Creating proxy_nginx-proxy_1 ...
Creating proxy_whoami_1 ...
Creating proxy_nginx-proxy_1
Creating proxy_nginx-proxy_1 ... done
Creating proxy_nginx-proxy-letsencrypt_1 ...
Creating proxy_nginx-proxy-letsencrypt_1 ... done
Attaching to proxy_whoami_1, proxy_nginx-proxy_1, proxy_nginx-proxy-letsencrypt_1
whoami_1                   | Listening on :8000
nginx-proxy_1              | WARNING: /etc/nginx/dhparam/dhparam.pem was not found. A pre-generated dhparam.pem will be used for now while a new one
nginx-proxy_1              | is being generated in the background.  Once the new dhparam.pem is in place, nginx will be reloaded.
nginx-proxy_1              | forego     | starting dockergen.1 on port 5000
nginx-proxy_1              | forego     | starting nginx.1 on port 5100
nginx-proxy_1              | Generating DH parameters, 2048 bit long safe prime, generator 2
nginx-proxy_1              | dockergen.1 | 2019/05/17 22:45:02 Generated '/etc/nginx/conf.d/default.conf' from 2 containers
nginx-proxy_1              | dockergen.1 | 2019/05/17 22:45:02 Running 'nginx -s reload'
nginx-proxy_1              | dockergen.1 | 2019/05/17 22:45:02 Watching docker events
nginx-proxy_1              | dockergen.1 | 2019/05/17 22:45:03 Contents of /etc/nginx/conf.d/default.conf did not change. Skipping notification 'nginx -s reload'
nginx-proxy_1              | dockergen.1 | 2019/05/17 22:45:03 Received event start for container 489e1597992a
nginx-proxy_1              | dockergen.1 | 2019/05/17 22:45:03 Contents of /etc/nginx/conf.d/default.conf did not change. Skipping notification 'nginx -s reload'
nginx-proxy-letsencrypt_1  | Generating a RSA private key
nginx-proxy-letsencrypt_1  | ...............................................................................................................++++
nginx-proxy-letsencrypt_1  | ...........++++
nginx-proxy-letsencrypt_1  | writing new private key to '/etc/nginx/certs/default.key.new'
nginx-proxy-letsencrypt_1  | -----
nginx-proxy-letsencrypt_1  | Info: a default key and certificate have been created at /etc/nginx/certs/default.key and /etc/nginx/certs/default.crt.
nginx-proxy-letsencrypt_1  | Info: Creating Diffie-Hellman group in the background.
nginx-proxy-letsencrypt_1  | A pre-generated Diffie-Hellman group will be used for now while the new one
nginx-proxy-letsencrypt_1  | is being created.
nginx-proxy-letsencrypt_1  | Generating DH parameters, 2048 bit long safe prime, generator 2
nginx-proxy-letsencrypt_1  | Reloading nginx proxy (8ebed2aadd1d9cd2355b1b132bd12f049c48f5433883cb8ad739d0aead66ce7d)...
nginx-proxy-letsencrypt_1  | 2019/05/17 22:45:08 Generated '/etc/nginx/conf.d/default.conf' from 3 containers
nginx-proxy-letsencrypt_1  | 2019/05/17 22:45:08 [notice] 39#39: signal process started
nginx-proxy-letsencrypt_1  | 2019/05/17 22:45:09 Generated '/app/letsencrypt_service_data' from 3 containers
nginx-proxy-letsencrypt_1  | 2019/05/17 22:45:09 Running '/app/signal_le_service'
nginx-proxy-letsencrypt_1  | 2019/05/17 22:45:09 Watching docker events
nginx-proxy-letsencrypt_1  | 2019/05/17 22:45:09 Contents of /app/letsencrypt_service_data did not change. Skipping notification '/app/signal_le_service'
nginx-proxy-letsencrypt_1  | /etc/nginx/certs/whoami.mp0.rbgapp.com /app
nginx-proxy-letsencrypt_1  | Reloading nginx proxy (8ebed2aadd1d9cd2355b1b132bd12f049c48f5433883cb8ad739d0aead66ce7d)...
nginx-proxy-letsencrypt_1  | 2019/05/17 22:45:11 Generated '/etc/nginx/conf.d/default.conf' from 3 containers
nginx-proxy-letsencrypt_1  | 2019/05/17 22:45:11 [notice] 59#59: signal process started
nginx-proxy-letsencrypt_1  | Creating/renewal whoami.mp0.rbgapp.com certificates... (whoami.mp0.rbgapp.com)
nginx-proxy-letsencrypt_1  | 2019-05-17 22:45:14,895:INFO:simp_le:1382: Generating new account key
nginx-proxy-letsencrypt_1  | 2019-05-17 22:45:16,768:WARNING:simp_le:1397: --email was not provided; ACME CA will have no way of contacting you.
nginx-proxy-letsencrypt_1  | 2019-05-17 22:45:17,059:INFO:simp_le:1407: By using simp_le, you implicitly agree to the CA's terms of service: https://letsencrypt.org/documents/LE-SA-v1.2-November-15-2017.pdf
nginx-proxy-letsencrypt_1  | 2019-05-17 22:45:17,592:INFO:simp_le:1479: Generating new certificate private key
nginx-proxy_1              | nginx.1    | whoami.mp0.rbgapp.com 66.133.109.36 - - [17/May/2019:22:45:17 +0000] "GET /.well-known/acme-challenge/rQmqc1p1GshAs4LHkazPdn0eJ5pQafkDWXBDzPgXM08 HTTP/1.1" 200 87 "-" "Mozilla/5.0 (compatible; Let's Encrypt validation server; +https://www.letsencrypt.org)"
nginx-proxy-letsencrypt_1  | 2019-05-17 22:45:20,825:INFO:simp_le:360: Saving account_key.json
nginx-proxy-letsencrypt_1  | 2019-05-17 22:45:20,826:INFO:simp_le:360: Saving key.pem
nginx-proxy-letsencrypt_1  | 2019-05-17 22:45:20,827:INFO:simp_le:360: Saving chain.pem
nginx-proxy-letsencrypt_1  | 2019-05-17 22:45:20,827:INFO:simp_le:360: Saving fullchain.pem
nginx-proxy-letsencrypt_1  | 2019-05-17 22:45:20,828:INFO:simp_le:360: Saving cert.pem
nginx-proxy-letsencrypt_1  | /app
nginx-proxy-letsencrypt_1  | Reloading nginx proxy (8ebed2aadd1d9cd2355b1b132bd12f049c48f5433883cb8ad739d0aead66ce7d)...
nginx-proxy-letsencrypt_1  | 2019/05/17 22:45:22 Generated '/etc/nginx/conf.d/default.conf' from 3 containers
nginx-proxy-letsencrypt_1  | Sleep for 3600s
```

And this is the certificate my browser gets, which looks like the default self-signed certificate and not the LetsEncrypt one:

![image](https://user-images.githubusercontent.com/816821/57960010-5d3ce480-78bb-11e9-8dfb-a6f37aefa5d7.png)

What am I missing here?

**Comment by Greek64** [comment]
The `nginx-proxy` uses a default self-signed certificate (which is auto generated during start-up if you don't provide one yourself) for every host it doesn't know or have a certificate for.

Your `letsencrypt-nginx-proxy-companion` does create a certificate, but does store it locally in its container, and the `nginx-proxy` has no way to access it (or know that it exists).  
Follow the [Step 2](https://github.com/JrCs/docker-letsencrypt-nginx-proxy-companion#step-2---letsencrypt-nginx-proxy-companion) of the description, so that you map the required volumes from nginx to the letsencrypt-companion.

**Comment by elifiner** [comment]
I tried multiple browsers as well as curl and I'm getting the same results. I'm also getting the same results when running the 3 steps from the readme directly, without docker-compose. Inside both containers the /etc/nginx/certs directory has a whoami.mp0.rbgapp.com subdirectory but it's empty. I'll poke some more and post an update.

**Comment by elifiner** [comment]
Looks like I needed to wait until the Diffie-Hellman group finished generating and Nginx was reloaded, the certificate became valid only after these lines appeared in the log:

```
nginx-proxy-letsencrypt_1  | This is going to take a long time
nginx-proxy-letsencrypt_1  | Info: Diffie-Hellman group creation complete, reloading nginx.
nginx-proxy-letsencrypt_1  | Reloading nginx proxy (0343ec173351d6f44784390af17f5956223691e55d227f0acf53edcc2400cb51)...
```

A log line saying "whois.mp0.rbgapp.com in now ready and serving HTTPS traffic" or something similar could be useful here but it's my bad for jumping the gun.

By the way, the DHPARAM_GENERATION environment variable mentioned in [nginx-proxy README](https://github.com/jwilder/nginx-proxy#ssl-support-using-letsencrypt) doesn't seem to have an effect. For my purposes being able to quickly spin up containers is more important than tighter security and it would be nice to have that option.

**Comment by Greek64** [comment]
> By the way, the DHPARAM_GENERATION environment variable mentioned in nginx-proxy README doesn't seem to have an effect.

It does, but not in the way you think.  
Originally both `nginx-proxy` and `letsencrypt-companion` were generating DHPARAM files. The `DHPARAM_GENERATION` is an attempt to prevent double generation, so only the `letsencrypt-companion` generates one.  
If you want to disable it completely, copy the pregenerated default DHPARAM file ("/app/dhparam.pem.default") into "/etc/nginx/certs/dhparam.pem".

But the thing is, that it should not be required to wait for the DHPARAM generation to complete before you can access the site, as a default DHPARAM file (the one mentioned above) should be used until the new one is generated.  
Maybe @buchdag can shed some light here as to why you have that behavior.

**Comment by moribellamy** [comment]
In case it helps...I had this same problem, but I downgraded to 1.9.1 (latest "stable" version) and things seem to work now. Had to delete all my docker volumes as well.

**Comment by buchdag** [comment]
I'll do some tests when I'll have some spare time but the expected behavior is that a default DH parameters file is used while another is generated in the background (which should not block anything), as described by @Greek64 

@elifiner you have to understand that `nginx-proxy` and this container aren't one and the same project by the same peoples. The latter is made to work with the former, but in substance it's "just" bash automation around an ACME client to have Let's Encrypt certificate automatically issued and managed. Even if they are meant to complement each other, you have to treat the two separate projects as entirely separate.

If `nginx-proxy` does not serve the correct certificate it means the configuration it generated (through `docker-gen`) is wrong. The only thing this container can actively do about `nginx-proxy` configuration is attempting to trigger its re-creation and reload the nginx process so that the new configuration is applied.

**Comment by buchdag** [comment]
@moribellamy by same problem do you mean you had to wait until the DH parameters file generation to get a valid certificate, even if it was already generated according to the logs ?

**Comment by buchdag** [comment]
As to why `nginx-proxy`'s `DHPARAM_GENERATION` might not work : https://github.com/jwilder/nginx-proxy/pull/913#issuecomment-478276705

**Comment by buchdag** [comment]
I think I found the issue : openssl erase the target file as soon as its start generating the DH parameters, while previous tests led me to believe that it overwrote the file only **after** successfully generating the new DH parameters.

This looks similar to https://github.com/jwilder/nginx-proxy/pull/1213 and the same solution is being applied and tested.

**Comment by buchdag** [comment]
This seems to have been caused by a change in Alpine's OpenSSL / LibreSSL : as @moribellamy pointed out, in `v1.9.1` and `v1.10` (both based on `alpine:3.8`) OpenSSL have the expected behaviour of not overwriting the target file until after the command completion.

I guess a DH parameters test unit is needed.

**Comment by dm17** [comment]
For those of us who want to use docker-compose up to launch our services with the letsencrypt/nginx companion: what is the recommended way to do that in a docker-compose.yaml? If we already have a docker-compose.yaml for our services, then could we use another docker-compose.yaml to encapsulate it, like this: https://stackoverflow.com/questions/33408118/inheritance-or-nesting-with-docker-compose

Typically, we'd want two docker-compose.yaml - one for dev and one for prod. The only difference would be nginx/letsencrypt it seems.
